language/stdlib/modules/event.mvir

```diff
@@ -0,0 +1,90 @@
+// This module is used for emitting events into the event_store.
+module Event {
+    import 0x0.AddressUtil;
+    import 0x0.BytearrayUtil;
+    import 0x0.Hash;
+    import 0x0.U64Util;
+
+    // A resource representing the counter used to generate uniqueness under each account. There won't be destructor for
+    // this resource to guarantee the uniqueness of the generated handle.
+    resource HandleIdGenerator {
+        // A monotonically increasing counter
+        counter: u64,
+    }
+
+    // A handle for a event such that:
+    // 1. Other modules can emit event to this handle.
+    // 2. Storage can use this handle to prove the total number of events that happened in the past.
+    resource Handle {
+        // Total number of events emitted to this event stream.
+        counter: u64
+        // A globally unique ID for this event stream.
+        guid: bytearray,
+    }
+
+    // Derive a fresh unique id by using sender's HandleIdGenerator. The generated bytearray is indeed unique because it
+    // was derived from the hash(sender's HandleIdGenerator || sender_address). This module guarantees that the
+    // HandleIdGenerator is only going to be monotonically increased and there's no way to revert it or destroy it. Thus
+    // such counter is going to give distinct value for each of the new event stream under each sender. And since we
+    // hash it with the sender's address, the result is guaranteed to be globally unique.
+    fresh_guid(): bytearray {
+        let counter: &mut R#Self.HandleIdGenerator;
+        let exists_counter: bool;
+        let count: &mut u64;
+        let count_bytes: bytearray;
+        let sender: address;
+        let sender_bytes: bytearray;
+        let preimage: bytearray;
+        let ret: bytearray;
+
+        sender = get_txn_sender();
+        exists_counter = exists<HandleIdGenerator>(copy(sender));
+        if (!move(exists_counter)) {
+            move_to_sender<HandleIdGenerator>(HandleIdGenerator { count: 0 });
+        }
+        counter = borrow_global<HandleIdGenerator>(move(sender));
+        count = &mut move(counter).counter;
+        *copy(count) = *copy(count) + 1;
```

Comment:
This means the first event handle created from an account will have count 1. Is that what we want? It might make sense to move this increment after the read of `count` at 50